File: modules/wem-api/src/main/java/com/enonic/wem/api/entity/Attachment.java
package com.enonic.wem.api.entity;


import com.enonic.wem.api.blob.BlobKey;

public final class Attachment
{
    private final String name;

    private final long size;

    private final String mimeType;

    private final BlobKey blobKey;

    private Attachment( final Builder builder )
    {
        this.name = builder.name;
        this.size = builder.size;
        this.mimeType = builder.mimeType;
        this.blobKey = builder.blobKey;
    }

    public String name()
    {
        return name;
    }

    public long size()
    {
        return size;
    }

    public String mimeType()
    {
        return mimeType;
    }

    public BlobKey blobKey()
    {
        return blobKey;
    }

    public static Builder newAttachment()
    {
        return new Builder();
    }

    public static final class Builder
    {
        private String name;

        private long size;

        private String mimeType;

        private BlobKey blobKey;

        public Builder name( final String value )
        {
            this.name = value;
            return this;
        }

        public Builder size( final long value )
        {
            this.size = value;
            return this;
        }

        public Builder mimeType( final String value )
        {
            this.mimeType = value;
            return this;
        }

        public Builder blobKey( final BlobKey value )
        {
            this.blobKey = value;
            return this;
        }

        public Attachment build()
        {
            return new Attachment( this );
        }
    }


}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/attachment/Attachment.ts
module api.content.attachment {

    export class Attachment implements api.Equitable {

        private name: AttachmentName;

        private label: string;

        private mimeType: string;

        private size: number;

        constructor(builder: AttachmentBuilder) {
            this.name = builder.name;
            this.label = builder.label;
            this.mimeType = builder.mimeType;
            this.size = builder.size;
        }

        getBinaryReference(): api.util.BinaryReference {
            return new api.util.BinaryReference(this.name.toString());
        }

        getName(): AttachmentName {
            return this.name;
        }

        getLabel(): string {
            return this.label;
        }

        getMimeType(): string {
            return this.mimeType;
        }

        getSize(): number {
            return this.size;
        }

        equals(o: api.Equitable): boolean {
            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, Attachment)) {
                return false;
            }

            var other = <Attachment>o;

            if (!api.ObjectHelper.equals(this.name, other.name)) {
                return false;
            }

            if (!api.ObjectHelper.stringEquals(this.label, other.label)) {
                return false;
            }

            if (!api.ObjectHelper.stringEquals(this.mimeType, other.mimeType)) {
                return false;
            }

            if (!api.ObjectHelper.numberEquals(this.size, other.size)) {
                return false;
            }

            return true;
        }

        toJson(): api.content.attachment.AttachmentJson {
            return {
                "name": this.getName().toString(),
                "label": this.getLabel(),
                "mimeType": this.getMimeType(),
                "size": this.getSize()
            };
        }

        public static create(): AttachmentBuilder {
            return new AttachmentBuilder();
        }

    }

    export class AttachmentBuilder {

        name: AttachmentName;

        label: string;

        mimeType: string;

        size: number;

        public fromJson(json: AttachmentJson): AttachmentBuilder {
            this.setName(new AttachmentName(json.name)).
                setLabel(json.label).
                setSize(json.size).
                setMimeType(json.mimeType);
            return this;
        }


        public setName(value: AttachmentName): AttachmentBuilder {
            this.name = value;
            return this;
        }

        public setLabel(value: string): AttachmentBuilder {
            this.label = value;
            return this;
        }

        public setMimeType(value: string): AttachmentBuilder {
            this.mimeType = value;
            return this;
        }

        public setSize(value: number): AttachmentBuilder {
            this.size = value;
            return this;
        }

        public build(): Attachment {
            return new Attachment(this);
        }
    }
}
